### PR TITLE
[VIT-7375] Apple Health Kit: add support for nutritional data

### PR DESCRIPTION
**Pull request by ItachiEU**
Apologies for the large PR, but in my defence most of the lines is just listing all the different nutrients in different forms 🙈

### CODE REVIEW

File: Sources/VitalCore/Core/Client/Data Models/Patches/QuantitySample.swift [hunk 1]
```diff
@@ -55,10 +59,33 @@ public struct LocalQuantitySample: Hashable, Encodable {
...
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(value)
+    hasher.combine(startDate)
+    hasher.combine(endDate)
+    hasher.combine(sourceBundle)
+    hasher.combine(productType)
+    hasher.combine(type)
+    hasher.combine(unit)
+    hasher.combine(metadata)
+  }
```
Comment (andersio) [inline Sources/VitalCore/Core/Client/Data Models/Patches/QuantitySample.swift:87]: You can remove these. All properties are Hashable, so the compiler should be able to synthesise these.

File: Sources/VitalHealthKit/HealthKit/Abstractions.swift [hunk 2]
```diff
@@ -145,6 +145,49 @@ extension VitalHealthKitStore {
...
+      HKQuantityType.quantityType(forIdentifier: .dietarySodium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryZinc)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryChromium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryCopper)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryIodine)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryManganese)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryMolybdenum)!,
+      HKQuantityType.quantityType(forIdentifier: .dietarySelenium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryWater)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryCaffeine)!:
```
Comment (andersio) [inline Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: We might want to exclude these for now, as they are being synced individually.

Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: That or we stop syncing these as individual resources, and remap them to `.nutrition(.caffeine)` and `.nutrition(.water)` to `.meal`.

**Review by andersio — APPROVED** [review]
LGTM overall but we need to resolve the caffiene/water/meal overlap.

File: Sources/VitalHealthKit/HealthKit/Abstractions.swift (hunk 2 above)
Comment (ItachiEU) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Well there is not much harm in polling them for meal as well, right?

Comment (ItachiEU) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Keeping the individual .water and .meal can be useful for apps that only track hydration for example.

And then polling them separately with intention to wrap them in a meal object on top of that.

Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Ehm… I need to double check if there is harm. We used to enforce 1 HKSampleType -> 1 VitalResource for change observation purposes. But I might have changed that recently when breaking out timeseries from `VitalResource.activity`.

Comment (ItachiEU) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Ah I see, let me know

Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: You will need to remove these two here. We already have dedicated switch cases for water & caffeine above. Add `.meal` in those two places  instead.

File: Sources/VitalHealthKit/HealthKit/Models/VitalResource+HealthKit.swift [hunk 3]
```diff
@@ -288,6 +337,50 @@ func observedSampleTypes() -> [[HKSampleType]] {
...
+      HKQuantityType.quantityType(forIdentifier: .dietarySodium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryZinc)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryChromium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryCopper)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryIodine)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryManganese)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryMolybdenum)!,
+      HKQuantityType.quantityType(forIdentifier: .dietarySelenium)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryWater)!,
+      HKQuantityType.quantityType(forIdentifier: .dietaryCaffeine)!
```
[the remark this answers is not in the document]
Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Models/VitalResource+HealthKit.swift:381]: ditto, we have declared water & caffeine already. Either keep these & remove the earlier ones, or vice versa.

File: Sources/VitalHealthKit/HealthKit/Abstractions.swift (hunk 2 above)
Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]:
```diff
      case HKSampleType.quantityType(forIdentifier: .dietaryWater)!:
-       return [.nutrition(.water)]
+       return [.meal, .nutrition(.water)]
      case HKSampleType.quantityType(forIdentifier: .dietaryCaffeine)!:
-       return [.nutrition(.caffeine)]
+       return [.meal, .nutrition(.caffeine)]
```

File: Sources/VitalHealthKit/HealthKit/Models/VitalResource+HealthKit.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Models/VitalResource+HealthKit.swift:381]:
```suggestion
```

Comment (andersio) [inline Sources/VitalHealthKit/HealthKit/Models/VitalResource+HealthKit.swift:381]:
```suggestion
      HKQuantityType.quantityType(forIdentifier: .dietarySelenium)!
```

File: Sources/VitalHealthKit/HealthKit/Abstractions.swift (hunk 2 above)
Comment (ItachiEU) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Thanks. Though I just realised that we don't even have place for caffeine in our Meal model.
Should we update it?
(and it looks like none of the 3 apps I have on ios update the water nutrient)

Comment (ItachiEU) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: wdyt?

Comment (andersio) [reply Sources/VitalHealthKit/HealthKit/Abstractions.swift:188]: Honestly I would just exclude these two from meal syncing. Caffeine is rarely part of food nutritional labels. Water is tracked separately.

File: Sources/VitalHealthKit/HealthKit/Abstractions.swift [hunk 4]
```diff
@@ -118,10 +118,10 @@ extension VitalHealthKitStore {
         return [.vitals(.heartRate)]
 
       case HKSampleType.quantityType(forIdentifier: .heartRateVariabilitySDNN)!:
-        return [.vitals(.heartRateVariability)]
+      return [.meal, .vitals(.heartRateVariability)]
 
       case HKSampleType.quantityType(forIdentifier: .dietaryWater)!:
-        return [.nutrition(.water)]
+      return [.meal, .nutrition(.water)]
```
Comment (andersio) [inline Sources/VitalHealthKit/HealthKit/Abstractions.swift:124]: remove `.meal`?

**Review by andersio — APPROVED** [review]

**Pull request merged**